### ISSUE TITLE
Implement Memento RFC7089 endpoints in replay

### ISSUE THREAD

**Issue by machawk1** [oduwsdl/ipwb#105]
Current ipwb replay scheme does not provide access to multiple copies of the same URI. We should provide `/memento/`, `/timemap/`, and `/timegate/` endpoints. 

- [X] `/memento/`
- [X] `/timemap/`
- [x] `/timegate/`
- [x] `/memento/*/`

**Comment by machawk1** [comment]
Basic TimeMap endpoint written in issue-105 branch. May need to look into a better way of retaining or at least propagating the scheme from the user's request, at least for rel=original.

**Comment by machawk1** [comment]
Memento endpoint resolution of embedded resources is implemented in https://github.com/oduwsdl/ipwb/commit/c3d068e0c5321cc9468b4a6e1af011067f0bcc58 in issue-88 branch, to be merged.

**Comment by machawk1** [comment]
A memento within the system ought to reporting a TimeMap in the Link header of the HTTP response, e.g., curl -I http://127.0.0.1:5000/20160305192247/cs.odu.edu/~salam/ should have an HTTP response header `Link: http://127.0.0.1:5000/timemap/cdxj/cs.odu.edu/~salam/`

I also currently get an error HTTP 500 trying to access the above TM URI w/ InterPlanetary Wayback 0.2017.05.13.2201

**Comment by machawk1** [comment]
Bumped this priority, as LANL is looking to use ipwb in some way but requires the TimeGate endpoint to be fully implemented.

**Comment by machawk1** [comment]
@ibnesayeed I would appreciate any suggestions you have on approaches in accomplishing this. I have a method to complete this in my head but you tend to give good feedback after it's been implemented, which sometimes requires restructuring.

**Comment by ibnesayeed** [comment]
Implementing TimeGate should not take long as some of the binary lookup and redirect logic is already in place for the `/memento` endpoint. We can reuse that function as it is, but before calling it, we will have to extract value of `Accept-Datetime` header, if present (otherwise default to current time), then convert it to the 14-digit format before calling `showMemento()`. We also need to set some additional response headers such as `Via` when `/timegate` endpoint is hit to advertise the fact that we negotiate the content via `Accept-Datetime` header on TimeGate endpoint.

**Comment by machawk1** [comment]
@ibnesayeed Check out dddb63dbda00f1ff690729539175da58e4399223, please.

**Comment by machawk1** [comment]
Current state:
```sh
$ ipwb index ipwb/samples/warcs/5mementos.warc | ipwb replay
```

```sh
$ curl http://localhost:5000/timemap/link/memento.us/
<http://memento.us/>; rel="original",
<http://localhost:5000/timemap/link/memento.us/>; rel="self timemap"; type="application/link-format",
<http://localhost:5000/timemap/cdxj/memento.us/>; rel="timemap"; type="application/cdxj+ors",
<http://localhost:5000/20130202100000/memento.us/>; rel="first memento"; datetime="Sat, 02 Feb 2013 10:00:00 GMT",
<http://localhost:5000/20140114100000/memento.us/>; rel="memento"; datetime="Tue, 14 Jan 2014 10:00:00 GMT",
<http://localhost:5000/20140115101500/memento.us/>; rel="memento"; datetime="Wed, 15 Jan 2014 10:15:00 GMT",
<http://localhost:5000/20161231110000/memento.us/>; rel="memento"; datetime="Sat, 31 Dec 2016 11:00:00 GMT",
<http://localhost:5000/20161231110001/memento.us/>; rel="last memento"; datetime="Sat, 31 Dec 2016 11:00:01 GMT"

$ curl -i -H "Accept-Datetime: Tue, 14 Jan 2014 10:01:00 EST" http://localhost:5000/timegate/memento.us/
HTTP/1.0 302 FOUND
Content-Type: text/html; charset=utf-8
Content-Length: 277
Location: http://localhost:5000/memento/20140114100000/memento.us/
Server: InterPlanetary Wayback Replay/0.2018.06.10.1938
Date: Thu, 21 Jun 2018 21:32:17 GMT

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>Redirecting...</title>
<h1>Redirecting...</h1>
<p>You should be redirected automatically to target URL: <a href="/memento/20140114100000/memento.us/">/memento/20140114100000/memento.us/</a>.  If not click the link.

$ curl -i http://localhost:5000/timegate/memento.us/
HTTP/1.0 302 FOUND
Content-Type: text/html; charset=utf-8
Content-Length: 277
Location: http://localhost:5000/memento/20161231110001/memento.us/
Server: InterPlanetary Wayback Replay/0.2018.06.10.1938
Date: Thu, 21 Jun 2018 21:33:05 GMT

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>Redirecting...</title>
<h1>Redirecting...</h1>
<p>You should be redirected automatically to target URL: <a href="/memento/20161231110001/memento.us/">/memento/20161231110001/memento.us/</a>.  If not click the link.
```

In the process of adding the URI-T and rel.

**Comment by machawk1** [comment]
Updated to add TG in Link header, Link TMs and CDXJ TMs in 80bbdd017e6df9ed7d04b487683f77bfe06fbbe4

```sh
$ curl -IL -H "Accept-Datetime: Tue, 14 Jan 2014 10:01:00 EST" http://localhost:5000/timegate/memento.us/
HTTP/1.0 302 FOUND
Content-Type: text/html; charset=utf-8
Content-Length: 277
Location: http://localhost:5000/memento/20140114100000/memento.us/
Server: InterPlanetary Wayback Replay/0.2018.06.10.1938
Date: Thu, 21 Jun 2018 22:02:03 GMT

HTTP/1.0 200 OK
Content-Type: text/html
Content-Length: 148
X-Archive-Orig-Server: nginx
X-Archive-Orig-Date: Mon, 30 Jan 2017 18:39:49 GMT
X-Archive-Orig-Connection: close
X-Archive-Orig-Vary: Accept-Encoding
Memento-Datetime: Tue, 14 Jan 2014 10:00:00 GMT
Link: <http://memento.us/>; rel="original", <http://localhost:5000/timemap/link/memento.us/>; rel="self timemap"; type="application/link-format", <http://localhost:5000/timemap/cdxj/memento.us/>; rel="timemap"; type="application/cdxj+ors", <http://localhost:5000/timegate/memento.us/>; rel="timegate"; <http://localhost:5000/20130202100000/memento.us/>; rel="first prev memento"; datetime="Sat, 02 Feb 2013 10:00:00 GMT", <http://localhost:5000/20140114100000/memento.us/>; rel="memento"; datetime="Tue, 14 Jan 2014 10:00:00 GMT", <http://localhost:5000/20140115101500/memento.us/>; rel="next memento"; datetime="Wed, 15 Jan 2014 10:15:00 GMT",  <http://localhost:5000/20161231110001/memento.us/>; rel="last memento"; datetime="Sat, 31 Dec 2016 11:00:01 GMT"
Server: InterPlanetary Wayback Replay/0.2018.06.10.1938
Date: Thu, 21 Jun 2018 22:02:03 GMT

$ curl -i http://localhost:5000/timemap/cdxj/memento.us/
HTTP/1.0 200 OK
Content-Type: application/cdxj+ors
Content-Length: 1115
Server: InterPlanetary Wayback Replay/0.2018.06.10.1938
Date: Thu, 21 Jun 2018 22:02:39 GMT

!context ["http://tools.ietf.org/html/rfc7089"]
!id {"uri": "http://localhost:5000/timemap/cdxj/memento.us/"}
!keys ["memento_datetime_YYYYMMDDhhmmss"]
!meta {"original_uri": "http://memento.us/"}
!meta {"timegate_uri": "http://localhost:5000/timegate/memento.us/"}
!meta {"timemap_uri": {"link_format": "http://localhost:5000/timemap/link/memento.us/", "cdxj_format": "http://localhost:5000/timemap/cdxj/memento.us/"}}
20130202100000 {"uri": "http://localhost:5000/20130202100000/memento.us/", "rel": "first memento", "datetime"="Sat, 02 Feb 2013 10:00:00 GMT"}
20140114100000 {"uri": "http://localhost:5000/20140114100000/memento.us/", "rel": "memento", "datetime"="Tue, 14 Jan 2014 10:00:00 GMT"}
20140115101500 {"uri": "http://localhost:5000/20140115101500/memento.us/", "rel": "memento", "datetime"="Wed, 15 Jan 2014 10:15:00 GMT"}
20161231110000 {"uri": "http://localhost:5000/20161231110000/memento.us/", "rel": "memento", "datetime"="Sat, 31 Dec 2016 11:00:00 GMT"}
20161231110001 {"uri": "http://localhost:5000/20161231110001/memento.us/", "rel": "last memento", "datetime"="Sat, 31 Dec 2016 11:00:01 GMT"}$ 

$ curl -i http://localhost:5000/timemap/link/memento.us/
HTTP/1.0 200 OK
Content-Type: application/link-format
Content-Length: 850
Server: InterPlanetary Wayback Replay/0.2018.06.10.1938
Date: Thu, 21 Jun 2018 22:03:01 GMT

<http://memento.us/>; rel="original",
<http://localhost:5000/timemap/link/memento.us/>; rel="self timemap"; type="application/link-format",
<http://localhost:5000/timemap/cdxj/memento.us/>; rel="timemap"; type="application/cdxj+ors",
<http://localhost:5000/timegate/memento.us/>; rel="timegate"; <http://localhost:5000/20130202100000/memento.us/>; rel="first memento"; datetime="Sat, 02 Feb 2013 10:00:00 GMT",
<http://localhost:5000/20140114100000/memento.us/>; rel="memento"; datetime="Tue, 14 Jan 2014 10:00:00 GMT",
<http://localhost:5000/20140115101500/memento.us/>; rel="memento"; datetime="Wed, 15 Jan 2014 10:15:00 GMT",
<http://localhost:5000/20161231110000/memento.us/>; rel="memento"; datetime="Sat, 31 Dec 2016 11:00:00 GMT",
<http://localhost:5000/20161231110001/memento.us/>; rel="last memento"; datetime="Sat, 31 Dec 2016 11:00:01 GMT"
```

**Comment by machawk1** [comment]
TimeGate endpoint comprehensively implemented in #391.
